### PR TITLE
Update to python 3

### PR DESCRIPTION
**Pull request by danielquinn**
This PR is basically an update of the tutorial to reference Python 3 with a few additions and clarifications along the way:

### Part 1

1. I felt that the last section surprised the user with passing multiple arguments to print(), so I added a section to prepare them for it.
2. **I also winged it on the OSX installation as I don't have a mac**.  I'm honestly not sure if this is what one has to do for Mac people, so some tips here would help.

### Part 2

1. I added a section on *"types & casting"* because the last exercise had code that wouldn't run without casting the input as an integer.

### Part 3

1. I swapped out the use of `range()` for `len()` since we haven't introduced users to lists at this stage.
2. I changed the name of the function that returns `My name is <name>` since it doesn't actually print anything.

### Notes

This tutorial has a gaping hole in it in that it doesn't cover lists or dictionaries, arguably core components of a basic Python background.  I would argue that this should precede what's currently part 3, but didn't want to start rewriting the whole tutorial on my first PR about a Python update :-)

If there's appetite for it though, I'm happy to add additional sections.  I'd probably end up heavily cannibalising [a list of exercises and tutorials I already wrote for people new to Python](https://github.com/danielquinn/Python-Django-for-PHP-Nerds).

Also, why is this pull request against `gh-pages` and not `master`?

### CODE REVIEW

**Comment by octopusinvitro** [comment]
Hi Daniel!
Thank you for your contribution!
To answer your question, this repo is basically a Jekyll site served automatically by GitHub pages, so it makes sense that gh-pages is the main branch and all PRs should be merged into it. We could even delete the master branch, as there is nothing special about it, it's just another branch, see for example http://oli.jp/2011/github-pages-workflow/#deleting-master

**Comment by danielquinn** [comment]
@octopusinvitro I figured it had something to do with how the Jekyll site is served.  I know one *can* delete master, I was just surprised by the divergence from convention is all.  Thanks for explaining though!

File: python/lesson1/tutorial.md [hunk 1]
```diff
@@ -66,65 +89,66 @@ Everybody's favourite pastime.
...
+    >>> (10 - 2) * 4
 
 You should have a different answer.
 
 Because of precedence rules, complex operations such as our first example can be quite confusing to read. If you find yourself writing more complex expressions, there is no harm in adding parentheses for clarity.
 
 ### Decimal Points
 
-Remember dividing 14 by 4? The result probably surprised you. Fortunately, there is a way to convince Python that you really do want the full answer, not just the whole number. We can do this by *starting* with numbers with decimal points:
+One of the things that tends to confuse newcomers to programming in general is the concept of *floating point numbers*.  Basically, numbers with decimal points tend to behave a little strangely when you're performing mathematical operations on them.  The reasons for this are complex and rooted in the nature of computing itself, so for now, let's just go with the assumption that weird things happen with decimal numbers and leave it at that.
```
Comment (lwm) [inline python/lesson1/tutorial.md]: I think the original was clearer but I think you are right to add something more. Can you shorten this down substantially to make your point more concrete?

File: python/lesson1/tutorial.md [hunk 2]
```diff
@@ -66,65 +89,66 @@ Everybody's favourite pastime.
...
+    >>> 10 / 3
 
-You may notice that there are still some limitations. For example, try:
+The answer should go on forever, but it doesn't.  Now try something a little more precision-sensitive:
 
-    10.0 / 3.0
+    >>> 1.000000000000001 * 8
 
-The answer should go on forever, but it doesn't. This is because there is only a finite amount of space to store the number, and it runs out after approximately 15 significant figures.
+Probably not what you'd expect right?  For now, you'll just have to accept this as a limitation, and later on you'll learn how other programmers work around it.
```
Comment (lwm) [inline python/lesson1/tutorial.md]: I am not sure it is constructive to leave this mystery here. Up to you if you want to change that.

File: python/lesson2/tutorial.md [hunk 3]
```diff
@@ -39,70 +39,102 @@ As well as numbers variables are able to store text, known in Python as strings.
...
+Sometimes you will need to use an apostrophe within a single quote, on occasions like this it is recommended to use "string escaping". This would look like:
 
 	message ='I\'m a string'
 
 Try storing a string within a variable without quotes, see what happens? Numbers do not require quotation marks, whereas they are mandatory for storing strings.
 
 Now store some strings in variables that contain apostrophes and some that do not.
 
-What happens when you store a number in a variable wrapped in quotes? 
+## Types & casting
```
Comment (lwm) [inline python/lesson2/tutorial.md]: Bravo. I think this is very useful.

File: python/lesson3/tutorial.md [hunk 4]
```diff
@@ -56,42 +56,40 @@ Now that we have defined a parameter, the function will raise an error if no par
...
     >>> print_name("Lisa")
     My name is Lisa
     >>> print_name()
     My name is Bart
 
-
 ## Different kinds of functions
 
-A function can be put into one of two categories. Fruitful functions, and void functions. So far we have only created void functions, these are functions that perform tasks but don't return a value. Fruitful functions return a value, a familiar example is the range function we used earlier that returned a list. Let's change our function from void to fruitful.
+A function can be put into one of two categories. "Fruitful" functions, and "void" functions. So far we have only created void functions, these are functions that perform tasks but don't return a value. Fruitful functions return a value, a familiar example is the `len()` function we used earlier that returned an integer. Let's change our function from void to fruitful.
```
Comment (lwm) [inline python/lesson3/tutorial.md]: Something doesn't feel right. Why not just slim this down with - Our function returns `None`, let's give it a return value. I don't think I've heard the 'fruitful' terminology being used before? Could be nitpicking ... ;)

File: python/lesson3/tutorial.md [hunk 5]
```diff
@@ -56,42 +56,40 @@ Now that we have defined a parameter, the function will raise an error if no par
...
+Python comes with a lot of built-in functions that are worth learning about, as they have been created specifically to perform the most common tasks that developers require. They are listed in the [documentation](https://docs.python.org/3/library/functions.html "Python 3 built in functions documentation"). To begin with, try using a few to see what they do, and how they behave with varying parameters. Once you have grasped how a few of them work, try writing your own functions that call the built in functions within them, like we did when we used the `print` function in our own custom `print_name()` function.
```
Comment (lwm) [inline python/lesson3/tutorial.md]: Nice. Could also recommend this useful reference: https://learnxinyminutes.com/docs/python3/

File: python/lesson3/tutorial.md [hunk 6]
```diff
@@ -56,42 +56,40 @@ Now that we have defined a parameter, the function will raise an error if no par
...
-    >>> def print_name(name='Bart'):
+    >>> def get_name_statement(name='Bart'):
     ...    return "My name is " + name
     
 Now the function returns a value that we can work with. We could store this in a value to be used later. Let's try storing it as a value, and using that value as a parameter in another `print_name()` function call.
 
-    >>> printed_name = print_name("Lisa")
-    >>> print print_name(printed_name)
+    >>> name_statement = get_name_statement("Lisa")
+    >>> print print_name(name_statement)
```
Comment (KimberleyCook) [inline python/lesson3/tutorial.md]: have you double checked that the `print_name` function still works at this point?

File: python/lesson1/tutorial.md (hunk 1 above)
Comment (danielquinn) [inline python/lesson1/tutorial.md]: I just stared at the paragraph for a good long time and can't seem to figure out how to shorten it nicely.  I'm open to suggestions though :-/

File: python/lesson1/tutorial.md (hunk 2 above)
Comment (danielquinn) [inline python/lesson1/tutorial.md]: To be honest, *I don't actually know why this happens*.  Just that it has something to do with how numbers are stored in memory, limitations of physics or something.  I've always just been happy to live with the mystery, but if you have an explanation we can put in here I'm all ears.

File: python/lesson2/tutorial.md (hunk 3 above)
Comment (danielquinn) [inline python/lesson2/tutorial.md]: Thanks!

File: python/lesson3/tutorial.md (hunk 4 above)
Comment (danielquinn) [inline python/lesson3/tutorial.md]: Yeah I didn't much like the original wording, but didn't want to start taking a hatchet to everything in my first PR.  I like your version, so I'll use that.

File: python/lesson3/tutorial.md (hunk 6 above)
Comment (danielquinn) [inline python/lesson3/tutorial.md]: Nooooope.  Oops.  I'll fix it now.

**Comment by danielquinn** [comment]
Alright I've amended the PR to include the requested changes -- all but one that I honestly couldn't think of better wording, but I'm happy to do whatever @lwm suggests on that one.

I was thinking that now that the installation is rather long, it might be a good idea to re-break it all up like so:

1. Installation
2. Integers, Floats, Strings, and Maths
3. Lists & Dictionaries *(doesn't exist yet)*
4. Variables
5. Functions

I was also thinking of introducing a hard line break at 79 characters, but didn't want to do that unless there's a standard of intentionally not wrapping the text.

Any of this can be done for this PR or added as a separate one.  I leave it up to the group to let me know what you want.

File: python/lesson1/tutorial.md (hunk 1 above)
Comment (lwm) [inline python/lesson1/tutorial.md]: Replace 'The reasons for this are complex and rooted in the nature of computing itself, so for now, let's just go with the assumption that weird things happen with decimal numbers and leave it at that.' with 'You'll figure it out as you go along.'

File: python/lesson1/tutorial.md (hunk 2 above)
Comment (lwm) [inline python/lesson1/tutorial.md]: Fair enough. Should be fine!

**Comment by lwm** [comment]
I've added my final comments (sorry for the delay!).

Go ahead and break that up then into the sections you outlined. LGTM.

I'd welcome the hard line break.

File: python/lesson1/tutorial.md (hunk 1 above)
Comment (KimberleyCook) [inline python/lesson1/tutorial.md]: hm I still not a fan of 'You'll figure it out as you go along'. It think we who should just end it at 'the assumption that weird things happen with decimal numbers' .

Comment (danielquinn) [inline python/lesson1/tutorial.md]: How 'bout *"..., so for now, let's just go with the understanding that weird things happen with decimal numbers."*?

Comment (KimberleyCook) [inline python/lesson1/tutorial.md]: Yes :)

Comment (KimberleyCook) [inline python/lesson1/tutorial.md]: when this is done I'll merge :D

**Comment by danielquinn** [comment]
Alright based on the feedback I see here I've got 3 things to do before it's merge-ready:

1. Fix the sentence about floating point math based on the comments above
2. Add the section on lists & dicts and break out the installation
3. Squash everything into one commit.

I'll try to do all of that tonight when I get home from work.

**Comment by danielquinn** [comment]
Alright, I've pushed a rather epic update.  The only real difference though is (a) the addition of a tutorial on lists, tuples, and dictionaries, and conforming everything to 79 characters.  I also made the one textual change requested.

**Comment by KimberleyCook** [comment]
I don't think you needed to split up that first tutorial. It's ok to have installation guide at the top of tutorial 1 then start playing with Strings, Integers and Floats. Either tutorial one should be renamed to installation as it currently isn't an 'Introduction to Python' as there is no tutorial in it just steps to install Python or the tutorial part of it needs to go back?

Also do we need installation guides for for Redhat/CentOS, Arch, or Gentoo ?

I do really appreciate the work that has been done on this but I feel it has gone a bit further then the initial intention.

**Comment by KimberleyCook** [comment]
I think renaming to 'Installation Guide' is a good route. If that could be changed I'll merge :)

**Comment by danielquinn** [comment]
Sorry about that, I thought it was agreed that I'd split up part 1, but if you'd prefer them to be one section, I can fix that.  As for the other Linuxes, I think it's a good idea to at least include RedHat/CentOS because it's so pointlessly complicated (nice job RedHat) and because it's entirely possible that someone will only have a company machine as a work environment and CentOS is very common in enterprise installations.  I added Arch & Gentoo out of a sense of completeness and because they're nice and short.  If you'd prefer though, I can toss out everything south of Debian/Ubuntu and recombine tutorials 1 & 2.  My feelings won't be hurt :-)

**Comment by KimberleyCook** [comment]
No I think keep as it's all really good but actually rename the first to Installation Guide rather than Introduction to Python

**Comment by danielquinn** [comment]
Title: updated!

**Comment by KimberleyCook** [comment]
@danielquinn Thank you so much for this epic update :D

**Comment by danielquinn** [comment]
No problem, it was fun :-)  If there's interest in more, (flow controls, classes, metaclasses, etc.) I'd be happy to add them.  There's also the question of whether we want to add stuff for frameworks like Django/Flask or if we would just point people to something like [DjangoGirls](https://djangogirls.org/).

**Comment by KimberleyCook** [comment]
Cool, I just did a PR updating the index file to link to your update. All up on http://tutorials.codebar.io/ . 

A python tutorial on classes I can imagine would be very useful. We have tried to stay away from frameworks but we could do something similar to the ruby one, where in the last tutorial (in the further reading section) we mention rails http://tutorials.codebar.io/ruby/lesson5/tutorial.html

**Comment by danielquinn** [comment]
Agreed.  There's no sense in going all-out and doing a Django tutorial when we can just point people somewhere.  I just wanted to get an idea of how comprehensive you wanted this to get.

One thing I should probably mention though: right now you've got the index page linking to `lesson2` as "Lesson 1", `lesson3` as "Lesson 2" and so on.  Maybe it could look something like this instead?

* Lesson 1: Installing Python
* Lesson 2: Strings, Integers and Floats
...

Either that, or we can rename the directories to start with `lesson0`?

**Comment by KimberleyCook** [comment]
Yes, was just amending that. Firstly I wanted to make sure I was linking to them all.

**Comment by lwm** [comment]
:rocket:

File: python/lesson1/tutorial.md [hunk 7]
```diff
@@ -1,135 +1,114 @@
...
+### Mac
 
-On Windows:
+Mac OSX ships with Python 2.7 which, while still a great language, is very
+outdated these days, so we need to upgrade:
+
+1. Make sure you're running an updated version of OSX.  You should at least
+   have OS 10.3, and most modern Macs are running much later than that anyway,
+   but it's worth noting this requirement here.
+2. Head to [the Python website](https://www.python.org/).
```
Comment (SamirTalwar) [inline python/lesson1/tutorial.md]: Is it worth suggesting Homebrew instead? I know a bunch of the students already have it.

File: python/lesson1/tutorial.md [hunk 8]
```diff
@@ -1,135 +1,114 @@
...
+    $ python
 
-### Further Reading
+And you should see something like this:
 
-There is a very good introductory article in [Google Developers Guide](https://developers.google.com/edu/python/introduction)
+    Python 3.5.2 (default, Aug  6 2016, 15:44:26) 
+    [GCC 5.4.0] on linux
+    Type "help", "copyright", "credits" or "license" for more information.
+    >>>
```
Comment (SamirTalwar) [inline python/lesson1/tutorial.md]: Worth explaining how to exit the REPL too, so people can *stop* after the end of the lesson.

(And re-explain how to start it in the next tutorial.)

**Comment by danielquinn** [comment]
I don't have a mac, so I don't know how to use Homebrew to install anything.  If you've got the instructions though, I'd agree that they should be in there.  Maybe under two headings: "Download from the web" and "Via Homebrew" or something like that.

Good call on the exiting the shell as well.  That's definitely a good thing to have in there.  Maybe now is a good time to mention that I really don't like the use of REPL here?  I've only ever seen that acronym used in Ruby land.  In Python, we generally just call it "the shell" no?

Anyway, this branch has already been merged and deleted, so changes here aren't really doable.  But if you want me to issue a new PR with the addition of the `exit` bit, I can do that later tonight.  The Homebrew stuff... that's on you :-)

**Comment by SamirTalwar** [comment]
I'm happy to make all the changes, including the `exit` bit, sometime this week. 😃

**Pull request merged**